snow/engine/snowman/bootstrap/interval/blocks.go

```diff
@@ -0,0 +1,237 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package interval
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/consensus/snowman"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/utils/set"
+	"github.com/ava-labs/avalanchego/utils/timer"
+)
+
+const (
+	batchWritePeriod      = 64
+	iteratorReleasePeriod = 1024
+	logPeriod             = 5 * time.Second
+)
+
+type Parser interface {
+	ParseBlock(context.Context, []byte) (snowman.Block, error)
+}
+
+// GetMissingBlockIDs returns the ID of the blocks that should be fetched to
+// attempt to make a single continuous range from
+// (lastAcceptedHeight, highestTrackedHeight].
+//
+// For example, if the tree currently contains heights [1, 4, 6, 7] and the
+// lastAcceptedHeight is 2, this function will return the IDs corresponding to
+// blocks [3, 5].
+func GetMissingBlockIDs(
+	ctx context.Context,
+	db database.KeyValueReader,
+	parser Parser,
+	tree *Tree,
+	lastAcceptedHeight uint64,
+) (set.Set[ids.ID], error) {
+	var (
+		missingBlocks     set.Set[ids.ID]
+		intervals         = tree.Flatten()
+		lastHeightToFetch = lastAcceptedHeight + 1
+	)
+	for _, i := range intervals {
+		if i.LowerBound <= lastHeightToFetch {
+			continue
+		}
+
+		blkBytes, err := GetBlock(db, i.LowerBound)
+		if err != nil {
+			return nil, err
+		}
+
+		blk, err := parser.ParseBlock(ctx, blkBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		parentID := blk.Parent()
+		missingBlocks.Add(parentID)
+	}
+	return missingBlocks, nil
+}
+
+// Add the block to the tree and return if the parent block should be fetched,
+// but wasn't desired before.
+func Add(
+	db database.KeyValueWriterDeleter,
+	tree *Tree,
+	lastAcceptedHeight uint64,
+	blk snowman.Block,
+) (bool, error) {
+	var (
+		height            = blk.Height()
+		lastHeightToFetch = lastAcceptedHeight + 1
+	)
+	if height < lastHeightToFetch || tree.Contains(height) {
+		return false, nil
+	}
+
+	blkBytes := blk.Bytes()
+	if err := PutBlock(db, height, blkBytes); err != nil {
+		return false, err
+	}
+
+	if err := tree.Add(db, height); err != nil {
+		return false, err
+	}
+
+	return height != lastHeightToFetch && !tree.Contains(height-1), nil
+}
+
+// Execute all the blocks tracked by the tree. If a block is in the tree but is
+// already accepted based on the lastAcceptedHeight, it will be removed from the
+// tree but not executed.
+//
+// Execute assumes that GetMissingBlockIDs would return an empty set.
+func Execute(
+	ctx context.Context,
+	log logging.Func,
```

Comment:
Taking in the log function rather than the full logger allows the caller to specify the log level.